File: test/stream_test.cpp
#include <iostream>
#include <string>
#include <mutex>
#include <condition_variable>
#include "dogears/dogears.h"
#include "test_utilites.h"

std::mutex mutex;
std::condition_variable cv;

int main(int argc, char* argv[]) {
    dogears::DogEars cape;

    auto passed = true;
    auto bufferNumber = 0;
    cape.beginStream<dogears::Normalized>([&](auto buffer) {
        passed &= checkBuffer(buffer, "Buffer " + std::to_string(bufferNumber));
        passed &= assertEqual(buffer.channels(), 4u, "Incorrect number of channels in buffer " + std::to_string(bufferNumber));

        if (bufferNumber == 10) {
            std::lock_guard<std::mutex> lock(mutex);
            cv.notify_all();
        }
        bufferNumber++;
    });

    std::unique_lock<std::mutex> lock(mutex);
    cv.wait(lock);
    cape.endStream();

    printTestResults(passed);
}

File: test/stream_test.py
import asyncio
import threading
from dogears import *
from test_utilites import *


with DogEars() as cape:
    event = threading.Event()
    passed = True
    bufferNumber = 0

    def callback(data):
        global passed, bufferNumber

        passed &= checkBuffer(data, "Buffer")
        passed &= assertEqual(data.shape[0], 4, "Incorrect number of channels")

        if bufferNumber == 10:
            event.set()
        bufferNumber += 1

    cape.beginStream(callback)

    event.wait()
    cape.endStream()
    printTestResults(passed)